src/management-system-v2/components/auth-can.tsx

```diff
@@ -42,17 +52,28 @@ export const AuthCan: FC<PropsWithChildren<AuthCanProps>> = ({
   const allow = useMemo(() => {
     if (status !== 'authenticated' || loadingState) return false;
 
-    const resources = Array.isArray(resource) ? resource : [resource];
-    const actions = Array.isArray(action) ? action : [action];
+    const selectedResources: (ResourceType | ReturnType<typeof toCaslResource>)[] = [];
+    for (const resource of resources) {
+      const selectedResource = props[resource];
+      if (!selectedResource) continue;
+
+      if (typeof selectedResource === 'boolean') selectedResources.push(resource);
+      else selectedResources.push(toCaslResource(resource, selectedResource));
+    }
+    if (selectedResources.length === 0) throw new Error('No resources were provided');
+
+    const actions: ResourceActionType[] = [];
+    for (const action of resourceAction) if (props[action]) actions.push(action);
+    if (selectedResources.length === 0) throw new Error('No actions were provided');
 
-    for (const r of resources) {
+    for (const r of selectedResources) {
       for (const a of actions) {
-        if (!ability.can(a, r)) return false;
+        if (!ability.can(a, r, { environmentId })) return false;
       }
     }
 
     return true;
-  }, [action, resource, ability, loadingState, status]);
+  }, [props]);
```

Comment:
Does this actually cache when the props don't change? Because I think the way JSX works is that the props object itself gets recreated on every render.